.github/workflows/reuse.yml

```diff
@@ -0,0 +1,254 @@
+      LEVIATHAN_DEBUG:
+        description: |
+          Use this parameter to modify the default debugging settings on test runs. To debug test runs when rebuilding, add new values to this parameter separated with commas. Know more about the possible values available for debugging: https://balena-os.github.io/leviathan/pages/Getting-Started/debugging.html Example value: dev: true, failFast: false
+        type: string
+        required: false
+    secrets:
+      BALENACLOUD_API_KEY:
+        required: false
+      BALENAMACHINE_API_KEY:
+        required: false
+
+jobs:
+  build:
+    runs-on: ubuntu-22.04
+    steps:
+    - name: Get Leviathan checked out
+      uses: actions/checkout@v4
+      with:
+        repository: balena-os/leviathan
+        submodules: true
+        # path: ./leviathan-workspace
+        # The branch, tag or SHA to checkout. When checking out the repository that
+        # triggered a workflow, this defaults to the reference or SHA for that event.
+        # Otherwise, uses the default branch.
+        ref: ${{ inputs.LEVIATHAN_REF }}
+
+    # - name: Set up Docker Buildx
+    #   uses: docker/setup-buildx-action@v3
+    # Don't think, I need this since docker-compose uses only buildkit which we enabled by default in Makefile
+
+    - name: Build it, and they might come
+      shell: bash
+      env:
+        WORKER_TYPE: ${{ inputs.WORKER_TYPE }}
+        # SECUREBOOT: ${{ inputs.SECUREBOOT }}
+      run: |
+        make config
+        make build
+        docker image ls -a
+
+    - name: Save the images
+      shell: bash
+      run: |
+        docker save --output /tmp/corelatest.tar leviathan_core:latest 
+        docker save --output /tmp/clientlatest.tar leviathan_client:latest 
+
+    - name: Upload Core artifact
+      uses: actions/upload-artifact@v3
+      with:
+        name: corelatest.tar
+        path: /tmp/corelatest.tar
+        compression-level: 7
+
+
+    - name: Upload Client artifact
+      uses: actions/upload-artifact@v3
+      with:
+        name: clientlatest.tar
+        path: /tmp/clientlatest.tar
+        compression-level: 7
+  
+
+  test:
+    runs-on: ubuntu-22.04
+    needs: build
+    steps:
+    - name: Download artifact
+      uses: actions/download-artifact@v3
+      with:
+        name: clientlatest.tar
+        path: /tmp
+
+    - name: Download artifact
+      uses: actions/download-artifact@v3
+      with:
+        name: corelatest.tar
+        path: /tmp
+
+    - name: Load image
+      shell: bash
+      run: |
+        docker image ls -a
+        docker load --input /tmp/corelatest.tar
+        docker load --input /tmp/clientlatest.tar
+        docker image ls -a
+
+    - name: Fetching code to be tested
+      uses: actions/checkout@v4
+      with:
+        submodules: true
+
+    - name: Get Leviathan checked out
+      uses: actions/checkout@v4
+      with:
+        repository: balena-os/leviathan
+        submodules: true
+        path: ./leviathan
```

Comment:
The device repo we checked out above already has a pinned copy of leviathan under `./layers/meta-balena/tests/leviathan` so we shouldn't check it out again at a different ref than was pinned.

We should only change the pinned submodule if someone has explicitly provided a leviathan ref as an input, and in that case we can do something like [this](https://github.com/balena-os/balena-yocto-scripts/blob/017d20d1165540a65835c658cf4e4a53779b9fa6/.github/workflows/yocto-build-deploy.yml#L219-L235).

Reply:
Yeah, I was a bit foggy on the whole git clone situations. Something to discuss definitely. I created a user-input version of leviathan checkout for testing.